### ISSUE TITLE
Encoding unknown parameters

### ISSUE THREAD

**Issue by kixelated** [moq-wg/moq-transport#219]
The SETUP/ANNOUNCE/SUBSCRIBE messages are encoded like this:

```
message_type:   varint
message_size:   varint
track_id:       varint
param_1_id:     varint
* param_1_size: varint
param_1_value:  bytes | varint
param_2_id:     varint
* param_2_size: varint
param_2_value:  bytes | varint
...

* ANNOUNCE/SUBSCRIBE do not state if each value has a size
```

Messages and parameters need a size if we support unknowns, otherwise it's impossible to skip them. However:

- **It's a waste**. This is particularly true for varints, as they have a variable length (1-8 bytes) and each of them must be prefixed with a varint size. This defeats the entire purpose of using the first two bits of a varint to encode the size.
- **It's a pain to encode**. An encoder needs to precompute the size of each message/param or use copy between temporary buffers. Other encodings like TLS and MP4 can seek backwards to populate the size later, but that's only possible with fixed size integers (not varints*) and is quite annoying. 
- **It's a pain to decode**. A decoder needs to use nested readers limited based on the size, or provide each `decode` method with the remaining length. It's a protocol error if any reader is done without 0 bytes remaining to read.
- **It prohibits streaming**. The required message length means you can't use stream encoders. We added a special case for 0 to mean undefined length, but that makes decoding more difficult and it prohibits future messages on the same stream.

**Comment by kixelated** [comment]
QUIC frames do not have sizes because it's illegal to use a frame/param not defined in the negotiated version. The only exception are transport parameters, as they are sent before a version is negotiated (similar to SETUP). It's efficient and a joy to encode and decode QUIC frames as a result. 

I propose that we copy this. We should explicitly _not_ support unknown messages and parameters after the CLIENT SETUP. The SERVER SETUP is prefixed with the chosen version so it doesn't need to support unknown parameters, but it could anyway for consistency.

Note that we can still support optional parameters. If we follow QUIC's example, the message type can contain flags on which fields are present. Or the zero value means a field is not present, for example the auth field would have size 0 when not used.

For example, here's what I would propose for SUBSCRIBE:
```
message_type: varint | group_present | object_present
track_id: varint
(group_seq: varint)
(object_seq: varint)
auth_size: varint
auth_value: bytes
```

It's lighter and significantly easier to encode/decode. A new version of the draft can add or remove fields; an implementation that supports multiple versions would use conditionals based on the version.

**Comment by fluffy** [comment]
One thought is to generalize this with a bitfield mask that indicated what options were in use.

**Comment by fluffy** [comment]
trvial nit but perhaps we should break out versioning it's own issue

**Comment by suhasHere** [comment]
I tend to agree that having Key/Value pair is pretty handy. I don't want us to start adding all options as bitmasks eventually. Mandatory fields go as the main fields, extensions/optionals go in key-value pairs.

**Comment by ianswett** [comment]
#255 provides a fix to this issue, but there was a question about whether something more complex (ie: structured headers) might eventually be necessary.

**Comment by martinduke** [comment]
Fixed by #255. Please open a new issue if the fix is unsatisfactory.
